### PR TITLE
Add globaldns commands

### PR DESCRIPTION
**Pull request by gitlawr**
Address issue: https://github.com/rancher/rancher/issues/17982

- add globaldns provider/entry commands
- add create/update/delete/list-members/add-member/delete-member
commands for provider/entry
- add add-project/delete-project commands for entry

Example command flows:
```
$ rancher globaldns provider ls
$ rancher globaldns provider create dnsp --type route53 --root-domain test.dm
$ rancher globaldns provider update dnsp --root-domain sub.test.dm
$ rancher globaldns provider delete p1 p2
$ rancher globaldns provider list-members p1
$ rancher globaldns provider add-member p1 user1 user2
$ rancher globaldns provider delete-member p1 user1 user2

$ rancher globaldns entry ls
$ rancher globaldns entry create --fqdn test.dm --provider p1
$ rancher globaldns entry update ENTRY_ID --fqdn test.dm
$ rancher globaldns entry delete ID1 ID2
$ rancher globaldns entry list-members ID
$ rancher globaldns entry add-member ID user1 user2
$ rancher globaldns entry delete-member ID user1 user2
$ rancher globaldns entry add-project ID p1 p2
$ rancher globaldns entry delete-project ID p1 p2
```

### CODE REVIEW

File: cmd/globaldns.go [hunk 1]
```diff
@@ -0,0 +1,865 @@
...
+		providerType = route53Provider
+	} else if provider.CloudflareProviderConfig != nil {
+		providerType = cloudflareProvider
+	} else if provider.AlidnsProviderConfig != nil {
+		providerType = alidnsProvider
+	}
+
+	switch providerType {
+	case route53Provider:
+		accessKey := os.Getenv(envAwsAccessKeyID)
```
Comment (orangedeng) [inline cmd/globaldns.go]: Should we have a way to get those parameters from cli flags instead of getting them from environment variable?

Comment (orangedeng) [reply cmd/globaldns.go]: like we can support --args AWS_ACCESS_KEY_ID=xxxxxxx --args AWS_SECRET_ACCESS_KEY=xxxx

File: cmd/globaldns.go [hunk 2]
```diff
@@ -0,0 +1,865 @@
...
+
+	if ctx.IsSet(argProvider) {
+		globalDNSProvider, err := searchForGlobalDNSProvider(c, ctx.String(argProvider))
+		if err != nil {
+			return err
+		}
+		update["providerId"] = globalDNSProvider.ID
+	}
+
+	if ctx.IsSet(argMultiClusterApp) {
```
Comment (orangedeng) [inline cmd/globaldns.go]: I think you need to validate that we can not support both `argProvider` and `argMultiClusterApp `  target in one update command.

Comment (gitlawr) [reply cmd/globaldns.go]: Why can't we do that?

Comment (orangedeng) [reply cmd/globaldns.go]: nvm, I see it now,

File: cmd/globaldns.go (hunk 1 above)
Comment (gitlawr) [reply cmd/globaldns.go]: Updated

**Comment by orangedeng** [comment]
LGTM

File: cmd/globaldns.go [hunk 3]
```diff
@@ -0,0 +1,931 @@
...
+	envAwsSecretKey          = "AWS_SECRET_ACCESS_KEY"
+	envCloudflareAPIEmail    = "CF_API_EMAIL"
+	envCloudflareAPIKey      = "CF_API_KEY"
+
+	argAwsAccessKey          = "aws-access-key"
+	argAwsSecretKey          = "aws-secret-key"
+	argCloudflareAPIEmail    = "cloudflare-api-email"
+	argCloudflareAPIKey      = "cloudflare-api-key"
+	argAlibabaCloudAccessKey = "alibabacloud-access-key-id"
+	argAlibabaCloudSecretKey = "alibabacloud-access-key-secret"
```
Comment (dramich) [inline cmd/globaldns.go]: Nit: Make alphabetical

File: cmd/globaldns.go [hunk 4]
```diff
@@ -0,0 +1,931 @@
...
+)
+
+const (
+	alidnsProvider     = "alidns"
+	cloudflareProvider = "cloudflare"
+	route53Provider    = "route53"
+
+	envAlibabaCloudAccessKey = "ALI_ACCESS_KEY_ID"
+	envAlibabaCloudSecretKey = "ALI_ACCESS_KEY_SECRET"
+	envAwsAccessKey          = "AWS_ACCESS_KEY_ID"
```
Comment (dramich) [inline cmd/globaldns.go]: Make these match the UI  e.g. `AWS_ACCESS_KEY`

File: cmd/globaldns.go [hunk 5]
```diff
@@ -0,0 +1,931 @@
...
+		return cli.ShowSubcommandHelp(ctx)
+	}
+
+	c, err := GetClient(ctx)
+	if err != nil {
+		return err
+	}
+
+	name := ctx.Args().First()
+	providerType := ctx.String(argType)
```
Comment (dramich) [inline cmd/globaldns.go]: strings.ToLower this

File: cmd/globaldns.go [hunk 6]
```diff
@@ -0,0 +1,931 @@
...
+							},
+							cli.StringFlag{
+								Name:  argRootDomain,
+								Usage: "Set root domain of a global DNS provider",
+							},
+							cli.StringSliceFlag{
+								Name:  argMember,
+								Usage: "Set members of the global DNS provider, can be used multiple times",
+							},
+							cli.StringFlag{
```
Comment (dramich) [inline cmd/globaldns.go]: Put the 6 key flags together outside the command creation then add the reference like it's done with formatFlag and quietFlag.

File: cmd/globaldns.go [hunk 7]
```diff
@@ -0,0 +1,931 @@
...
+}
+
+func getEntryTarget(c *cliclient.MasterClient, clusterCache map[string]managementClient.Cluster, projectCache map[string]managementClient.Project, entry *managementClient.GlobalDNS) (string, error) {
+	target := ""
+	if entry.MultiClusterAppID != "" {
+		_, app, err := searchForMcapp(c, entry.MultiClusterAppID)
+		if err != nil {
+			return "", err
+		}
+		target = fmt.Sprintf("Multi Cluster App %s", app.Name)
```
Comment (dramich) [inline cmd/globaldns.go]: Update to `Multi-Cluster App: %s`

File: cmd/globaldns.go [hunk 8]
```diff
@@ -0,0 +1,931 @@
...
+								Usage: "Set projects as the target to which a global DNS entry resolves, can be used multiple times",
+							},
+							cli.StringSliceFlag{
+								Name:  argMember,
+								Usage: "Set members of a global DNS entry, can be used multiple times",
+							},
+						},
+					},
+					{
+						Name:      "update",
```
Comment (dramich) [inline cmd/globaldns.go]: Why doesn't update support projects as a flag as well? Keeping the separate command for projects is fine.

File: cmd/globaldns.go [hunk 9]
```diff
@@ -0,0 +1,931 @@
...
+		return err
+	}
+
+	update := make(map[string]interface{})
+
+	if ctx.IsSet(argRootDomain) {
+		update["rootDomain"] = ctx.String(argRootDomain)
+	}
+
+	providerType := ""
```
Comment (dramich) [inline cmd/globaldns.go]: var providerType string

File: cmd/globaldns.go [hunk 10]
```diff
@@ -0,0 +1,931 @@
...
+			ID:     entry.ID,
+			Entry:  entry,
+			Target: target,
+		})
+	}
+	return writer.Err()
+}
+
+func getEntryTarget(c *cliclient.MasterClient, clusterCache map[string]managementClient.Cluster, projectCache map[string]managementClient.Project, entry *managementClient.GlobalDNS) (string, error) {
+	target := ""
```
Comment (dramich) [inline cmd/globaldns.go]: var target string

File: cmd/globaldns.go [hunk 11]
```diff
@@ -0,0 +1,931 @@
...
+	}
+
+	update := make(map[string]interface{})
+
+	if ctx.IsSet(argRootDomain) {
+		update["rootDomain"] = ctx.String(argRootDomain)
+	}
+
+	providerType := ""
+	if provider.Route53ProviderConfig != nil {
```
Comment (dramich) [inline cmd/globaldns.go]: Pull the switch up into these if blocks with an `else` for the error, `providerType` isn't used for anything else so no reason to check the same thing twice.

File: cmd/globaldns.go (hunk 8 above)
Comment (gitlawr) [reply cmd/globaldns.go]: Project target is `noupdate` fields in the API and it needs to be done via addProjects/removeProjects actions

File: cmd/globaldns.go (hunk 4 above)
Comment (gitlawr) [reply cmd/globaldns.go]: Where can I see `AWS_ACCESS_KEY` in the UI? I use this because it is commonly used by aws cli/sdk

**Comment by gitlawr** [comment]
@dramich Updated

File: cmd/globaldns.go (hunk 4 above)
Comment (dramich) [reply cmd/globaldns.go]: That makes sense, if that is how the aws cli works leave it.

**Comment by gitlawr** [comment]
Squashed

**Pull request merged**